napari/_vispy/visuals/volume.py

```diff
@@ -171,6 +95,127 @@
 }
 """
 
+FAST_GRADIENT_DEFINITION = """
+vec3 calculateGradient(vec3 loc, vec3 step) {
+    // calculate gradient within the volume by finite differences
+
+    vec3 G = vec3(0.0);
+
+    float prev;
+    float next;
+    int in_bounds;
+    float curr = colorToVal($get_data(loc));
+
+    for (int i=0; i<3; i++) {
+        vec3 ax_step = vec3(0.0);
+        ax_step[i] = step[i];
+
+        vec3 prev_loc = loc - ax_step;
+        if (u_clamp_at_border || (prev_loc[i] >= 0.0 && prev_loc[i] <= 1.0)) {
+            prev = colorToVal($get_data(prev_loc));
+        } else {
+            prev = categorical_bg_value;
+        }
+
+        vec3 next_loc = loc + ax_step;
+        if (u_clamp_at_border || (next_loc[i] >= 0.0 && next_loc[i] <= 1.0)) {
+            next = colorToVal($get_data(next_loc));
+        } else {
+            next = categorical_bg_value;
+        }
+
+        // add to the gradient where the adjacent voxels are both background
+        // to fix dim pixels due to poor normal estimation
+        G[i] = next - prev + (next - curr) * 2.0 * detectAdjacentBackground(prev, next);
+    }
+
+    return G;
+}
+"""
+
+SMOOTH_GRADIENT_DEFINITION = """
+vec3 calculateGradient(vec3 loc, vec3 step) {
+    // calculate gradient within the volume by finite differences
+    // using a 3D sobel-feldman convolution kernel
+
+    // the kernel here is a 3x3 cube, centered on the sample at `loc`
+    // the kernel for G.z looks like this:
+
+    // [ +1 +2 +1 ]
+    // [ +2 +4 +2 ]    <-- "loc - step.z" is in the center
+    // [ +1 +2 +1 ]
+
+    // [  0  0  0 ]
+    // [  0  0  0 ]    <-- "loc" is in the center
+    // [  0  0  0 ]
+
+    // [ -1 -2 -1 ]
+    // [ -2 -4 -2 ]    <-- "loc + step.z" is in the center
+    // [ -1 -2 -1 ]
+
+    // kernels for G.x and G.y similar, but transposed
+    // see https://en.wikipedia.org/wiki/Sobel_operator#Extension_to_other_dimensions
+
+    vec3 G = vec3(0.0);
+    float curr = 0.0;
+    // next and prev are the directly adjacent values along x, y, and z
+    vec3 next = vec3(0.0);
+    vec3 prev = vec3(0.0);
+
+    float val;
+    bool is_on_border = false;
+    for (int i=-1; i <= 1; i++) {
+        for (int j=-1; j <= 1; j++) {
+            for (int k=-1; k <= 1; k++) {
+                if (is_on_border && (i != 0 && j != 0 && k != 0)) {
+                    // we only care about on-axis values if we are on a border
+                    continue;
+                }
+                vec3 sample_loc = loc + vec3(i, j, k) * step;
+                bool is_in_bounds = all(greaterThanEqual(sample_loc, vec3(0.0)))
+                    && all(lessThanEqual(sample_loc, vec3(1.0)));
+
+                if (is_in_bounds || u_clamp_at_border) {
+                    val = colorToVal($get_data(sample_loc));
+                } else {
+                    val = categorical_bg_value;
+                }
```

Comment:
This is checking whether the sample will be in-bounds for texture sampling. Texture coordinates range from 0 to 1 (inclusive).

If it's in bounds or if we want the default behavior (clamping), we sample the texture. Otherwise if it's out of bounds, assume the background value.

Reply:
One of the concerns re making it configurable was the performance with the if statement. (That is, if the shader was slower with the if-statement, then we would want to be able to configure it and swap it out for the old shader.) I was convinced by @psobolewskiPhD that my visual reasoning for wanting configurability was flawed, so if we can make the shader simpler and/or more performant by dropping the u_clamp_at_border, I'd probably support that... (Does `$get_data(sample_loc)` simply return `categorical_bg_value` if outside the bounds anyway?)

Reply:
I kind of like leaving the configurability in as a private API - it makes it easier to add as a public config if there are requests.

I'm not too concerned about the performance, especially given the benchmark results you've seen. I'm not really a shader expert, but the rule-of-thumb I've heard about conditionals is something like you can expect performance as if both branches were executed. The check here should be cheap, and the alternate branch is also cheap.

> Does `$get_data(sample_loc)` simply return `categorical_bg_value` if outside the bounds anyway?

No, the opposite - `$get_data(sample_loc)` will clamp to the edges of the texture.

Reply:
Ah — I am remembering now [edge value sampling](https://www.khronos.org/opengl/wiki/Sampler_Object#Edge_value_sampling). Would changing the mode to GL_CLAMP_TO_BORDER (and is that easy to do in VisPy) be the more canonical way to address this?

Reply:
The above comment is more for discussion than as a TODO, or if anything it is a potential future TODO — I am happy with the code and it looks like there is no performance hit so let's get this in! 🚀

Reply:
> Would changing the mode to GL_CLAMP_TO_BORDER (and is that easy to do in VisPy) be the more canonical way to address this?

Yes! I just didn't see an easy way to do this in vispy.